python-sdk/nuscenes/map_expansion/arcline_path_utils.py

```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
     return discretization
 
 
-def discretize_lane(arcline_list: List[Dict[str, Any]],
+def discretize_lane(lane: List[ArcLinePath],
                     resolution_meters: float) -> List[Tuple[float, float, float]]:
     """
     Discretizes a lane and returns list of all the poses alone the lane.
-    :param arcline_list: Lanes are represented as a list of arcline_3_paths.
+    :param lane: Lanes are represented as a list of arcline_3_paths.
     :param resolution_meters: How finely to discretize the lane. Smaller values ensure turning
         lanes are properly represented.
     :return: List of pose tuples along the lane.
     """
 
     pose_list = []
-    for arcline_path in arcline_list:
-        poses = discretize(arcline_path, resolution_meters)
+    for path in lane:
+        poses = discretize(path, resolution_meters)
         for pose in poses:
             pose_list.append(pose)
     return pose_list
+
+
+def length_of_lane(lane: List[ArcLinePath]) -> float:
+    """
+    Calculates the length of a lane
+    :param lane: Lane
+    :return: Length of lane
+    """
+
+    return sum(sum(path['segment_length']) for path in lane)
+
+
+def project_pose_to_lane(pose: Pose, lane: List[ArcLinePath]) -> Tuple[Pose, float]:
+    """
+    Find the closest pose on a lane to a query pose and additionally return the
+    distance along the lane for this pose. Note that this function does
+    not take the heading of the query pose into account.
+    :param pose: Query pose
+    :param lane: Will find the closest pose on this lane
+    :return: Tuple of the closest pose and the distance along the lane
+    """
+
+    discretized_lane = discretize_lane(lane, resolution_meters=0.5)
+
+    xy_points = np.array(discretized_lane)[:, :2]
+    closest_pose_index = np.linalg.norm(xy_points - pose[:2], axis=1).argmin()
+
+    closest_pose = discretized_lane[closest_pose_index]
+    distance_along_lane = closest_pose_index * 0.5
+    return closest_pose, distance_along_lane
+
+
+def _find_index(distance_along_lane: float, lengths: List[float]) -> int:
+    """
+    Helper function for finding of path along lane corresponding to the distance_along_lane
+    :param distance_along_lane: Distance along the lane
+    :length: Cumulative distance at each end point along the paths in the lane
+    :return: Index of path
+    """
+
+    if len(lengths) == 1:
+        return 0
+    else:
+        return min(index for index, length in enumerate(lengths) if distance_along_lane <= length)
+
+
+def get_curvature_at_distance_along_lane(distance_along_lane: float, lane: List[ArcLinePath]) -> float:
+    """
+    Computes the unsigned curvature at a distance along a lane
+    :param distance_along_lane: Distance along the lane to calculate the curvature at
+    :param lane: Lane to query
+    :return: Curvature, always non negative.
+    """
+
+    segment_lengths = np.cumsum([sum(path['segment_length']) for path in lane])
```

Comment:
Is the name consistent with what you are computing? It is the length from start to end of each segment.

Reply:
You're right! Changed the name to total_length_at_segment